src/lib/bcmp/bcmp.cpp

```diff
@@ -488,3 +551,134 @@ void bcmp_init(struct netif* netif, NvmPartition * dfu_partition, Configuration*
+  bool rval = false;
+  configASSERT(message);
+  bcmp_element_t *element_to_delete = NULL;
+  if (_ctx.messages_list == message) {
+    element_to_delete = _ctx.messages_list;
+    _ctx.messages_list = _ctx.messages_list->next;
+  } else {
+    bcmp_element_t *current = _ctx.messages_list;
+    while (current && current->next != message) {
+      current = current->next;
+    }
+    if (current) {
+      element_to_delete = current->next;
+      current->next = current->next->next;
+    }
+  }
+  if (element_to_delete) {
+    vPortFree(element_to_delete);
+    rval = true;
+  }
+  return rval;
+}
+
+static bcmp_element_t *_message_list_create_element(uint16_t seq_num, uint16_t type, uint32_t timeout_ms) {
+  bcmp_element_t *element = static_cast<bcmp_element_t *>(pvPortMalloc(sizeof(bcmp_element_t)));
+  configASSERT(element);
+  element->seq_num = seq_num;
+  element->type = type;
+  element->send_timestamp_ms = pdTICKS_TO_MS(xTaskGetTickCount());
+  element->timeout_ms = timeout_ms;
+  element->next = NULL;
+  return element;
+}
+
+static void _message_list_timer_expiry_cb(void *arg) {
+  (void) arg;
+  if (xSemaphoreTake(_ctx.messages_list_mutex, pdMS_TO_TICKS(DEFAULT_MESSAGE_TIMEOUT_MS)) == pdPASS) {
+    bcmp_element_t *current = _ctx.messages_list;
+    while (current) {
+      if (pdTICKS_TO_MS(xTaskGetTickCount()) - current->send_timestamp_ms > current->timeout_ms) {
+        printf("BCMP message with seq_num %d timed out\n", current->seq_num);
+        _message_list_remove_message(current);
+        current = _ctx.messages_list;
+        continue;
+      }
+      current = current->next;
+    }
+    xSemaphoreGive(_ctx.messages_list_mutex);
+  }
+}
+
+static void _message_list_timer_callback(TimerHandle_t tmr){
+  (void) tmr;
+  timer_callback_handler_send_cb(_message_list_timer_expiry_cb, tmr, 0);
+}
+
+static bcmp_element_t *_message_list_find_message(uint16_t seq_num) {
+  bcmp_element_t *rval = NULL;
+  if (xSemaphoreTake(_ctx.messages_list_mutex, pdMS_TO_TICKS(DEFAULT_MESSAGE_TIMEOUT_MS)) == pdPASS) {
+    bcmp_element_t *current = _ctx.messages_list;
+    while (current) {
+      if (current->seq_num == seq_num) {
+        printf("BCMP message with seq_num %d found\n", current->seq_num);
+        rval = current;
+        break;
+      }
+      current = current->next;
+    }
+    xSemaphoreGive(_ctx.messages_list_mutex);
+  }
+  return rval;
+}
+
+static bool _message_is_sequenced_reply(uint16_t type) {
+  bool rval = false;
+  if (type == BCMP_CONFIG_VALUE ||
+      type == BCMP_CONFIG_STATUS_RESPONSE ||
+      type == BCMP_CONFIG_DELETE_RESPONSE ||
+      type == BCMP_DFU_PAYLOAD_REQ ||
+      type == BCMP_DFU_PAYLOAD ||
+      type == BCMP_DFU_END ||
+      type == BCMP_DFU_ACK ||
+      type == BCMP_DFU_ABORT ||
+      type == BCMP_DFU_HEARTBEAT ||
+      type == BCMP_DFU_REBOOT_REQ ||
+      type == BCMP_DFU_REBOOT ||
+      type == BCMP_DFU_BOOT_COMPLETE ||
+      type == BCMP_ECHO_REPLY) {
```

Comment:
I'm not sure if all of these are replies, especially w.r.g. to the DFU messages.

Reply:
Woops I see I already left in the DFU_REBOOT_REQ - I'll remove it!

Reply:
You can ignore all the DFU stuff for now since it's out of scope for this effort! (our just leave a TODO to audit the messages later

Reply:
right! I'll leave them out so they keep working as normal!